Clean-up implementation. Rewrite in typescript.

paulmillr opened the pull request:
Feels much simpler now.

## TODO

- [x] Justification for removing unsafe functions
- [x] Target ES2015 (We've received complaints before on other repos when targeting anything higher.)
- [x] Move tests to TS
- [ ] Move coverage to TS
- [x] Move linter to TS
- [x] Add git diff check in travis to prevent future PRs from modifying TS and JS only (npm run build will cause the TS to overwrite the JS, and if a git diff is born, that means the PR didn't modify both, only one)
- [x] Remove var (If we're going to go to TypeScript)

lukechilds commented on index.js:
```diff
@@ -1,182 +1,123 @@
+const ALPHABET_MAP = [...ALPHABET].reduce((res, el, i) => ({ ...res, [el]: i }), {});
+const GENERATORS = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];
+function polymodStep(pre) {
+    var b = pre >> 25;
```
Was use of `var` here intentional?

lukechilds commented on index.js:
```diff
@@ -1,182 +1,123 @@
+        throw new Error('Invalid hash length');
+    let chk = prefixChk(prefix), data = [];
+    for (let i = 0; i < dataChars.length; i++) {
+        var v = ALPHABET_MAP[dataChars[i]];
```
Was use of `var` here intentional?

lukechilds commented on index.ts:
```diff
@@ -0,0 +1,138 @@
+const GENERATORS = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];
+
+function polymodStep(pre: number): number {
+  var b = pre >> 25;
```
Was use of `var` here intentional?

paulmillr replied:
no

lukechilds commented on index.ts:
```diff
@@ -0,0 +1,138 @@
+  let chk = prefixChk(prefix),
+    data = [];
+  for (let i = 0; i < dataChars.length; i++) {
+    var v = ALPHABET_MAP[dataChars[i]];
```
Was use of `var` here intentional?

paulmillr commented:
Fixed `var` issue.

paulmillr commented:
The linter is obviously failing now because output JS cannot be customized from typescript compiler. I don't think this is a big deal.

Typescript code conforms to `prettier` linter, we can use it.

junderw commented:
coverage and standard are failing.

IMO, if we're going to include TS files, tests should be TS, coverage should be based on TS, standardized linting (prettier maybe?) checks should be against the TS. git-diffs should be included in Travis to prevent a Pull Request from modifying only the JS or only the TS.

etc. etc.

Also, from an implementation standpoint, you removed functions around without much justification.

A quick perusal had error messages such as "Not a 5-bit byte" when a byte only refers to 8-bit words, so the only term to refer to a member of an array of numbers between 0 and 31 is a "5-bit word".

If you're willing to commit to this I'm willing to review it, but for right now it is too much of a breaking change with too little justification other than "Feels simpler."

---

I have edited your main comment with a simple TODO list off the top of my head.

paulmillr commented:
We could move to GitHub actions after that's merged. I can provide a PR.

paulmillr commented:
@junderw 

>Justification for removing unsafe functions

They were never documented. What is the purpose? They complicate API.

junderw commented:
> They were never documented. What is the purpose? They complicate API.

unsafe functions never throw Errors and instead return undefined. Removing them is a breaking change.

paulmillr commented:
ok i've added them back but that still feels like unnecessary complication. JS folks use try-catch all the time, I don't see a point in doing this inside the lib.

added prettier.

junderw commented:
Here's what we do for git diff on bitcoinjs-lib:

https://github.com/bitcoinjs/bitcoinjs-lib/blob/91bb1e91aed97c4a31476c576d1810baeabdc055/package.json#L28

paulmillr commented:
@junderw coverage is now 100%.

I don't know how to move coverage to typescript. Do you think we could get away without it for now?

paulmillr commented:
All other parts are done. Feel free to merge.

paulmillr commented:
https://travis-ci.org/github/bitcoinjs/bech32/builds/701692127

junderw commented:
Thanks a lot. I think this is to a point where any fixes will be small, so if you can keep this pull request open to modifications from maintainers I'll just fix any minor things myself.

One larger conceptual thing I noticed that I might fix was:

1. Use of Uint8Array for return types is a breaking change. For parameter types it's a breaking change for typescript users.
2. Uint8Array doesn't make sense for 5-bit words. We used `ArrayLike<number>` for inputs (To allow for Buffer, Uint8Array, or whatever as an input) and `number[]` for outputs.

The rest will be mostly lint/format/coverage etc. changes.

paulmillr commented:
Using Uint8Array is an industry-standard way of working with binary data in javascript. Most cryptographic libraries work with Uint8Arrays. Browser crypto APIs use Uint8Arrays.

Buffer itself is a legacy node.js-only version of Uint8Array and is in fact a subclass of it. You can check this by doing: `Buffer.from([]) instanceof Uint8Array // => true`. So, using Uint8Array in type signatures would allow Buffers through. If you want to amend types to add `number[]`, i'd suggest to make it `Uint8Array | number[]`.

Maybe we could bump major version to ensure breaking changes are correctly reflected in semver.

junderw commented:
Yes. I am aware of `Buffer.from([]) instanceof Uint8Array` (Although this does not work in all environments. Recently I have had trouble caused by this assumption when we upgraded jest testing suite version.)

However, 5-bit words are not bytes. If someone sees the Uint8Array type and thinks "oh, these are 8-bit bytes inside each index of this Uint8Array. otherwise, why would they use Uint8Array?" they could possibly be misled. Also, doing things that many people do with byte arrays, ie. hashing, encoding into hex, etc. should not be done with 5-bit word arrays, and to discourage that, keeping it a number[] in lieu of Uint5Array (which to my knowledge doesn't exist...) is better in my opinion.

paulmillr commented:
It's fine if you want users to be able to use number[].

I think it's a big mistake to not allow Uint8Arrays directly. I'm using them, and most people are using them. I'm using them with bech32 code. They are standard across all libraries.

junderw commented:
Why would you want to put 5-bit words in a Uint8Array?

Each index of a Uint8Array represents values from `00000000` to `11111111` but if we only pass in values from `00000000` to `00011111` and someone decides to hash them, or whatnot, there are 3 extra padding bits for each byte which makes no sense.

In order to make sure that people using TypeScript do not think the 5-bit words are "bytes" it is not a good idea to use Uint8Array.

`Arraylike<number>` for parameters will accept Buffer, Uint8Array, number[] and is more desirable.

Return values should be stricter:

1. Return values of 5-bit word arrays should not be Uint8Array.
2. Return values of 8-bit word ("byte") arrays CAN be Uint8Array if we decide to break the API.

I am not quite sure which part you are talking about. I think our communication isn't working properly. What do you think I am suggesting?

I don't think I am suggesting anything outrageous here...

Here, this is the matrix of what I think the values should be, change it to what you think it should be and reply.

|                              |5-bit                           |8-bit                           |
|-------------------------|----------------------------|-----------------------------|
|input param           |`ArrayLike<number>`|`ArrayLike<number>`|
|output return value| `number[]`                | `Uint8Array`              |

You can copy-paste this and change it:

```
|                   |5-bit              |8-bit              |
|-------------------|-------------------|-------------------|
|input param        |`ArrayLike<number>`|`ArrayLike<number>`|
|output return value|`number[]`         |`Uint8Array`       |
```

paulmillr commented:
> Why would you want to put 5-bit words in a Uint8Array?
Uint8Array is the closest thing to Uint5Array. 

A number:

- can be more than 53-bit
- can be signed
- can be float
- can be NaN

> In order to make sure that people using TypeScript do not think the 5-bit words are "bytes" it is not a good idea to use Uint8Array.
There are way more incorrect things people can do with a number when compared to u8, and it also tells user limits about what to expect from output without looking at code.

> Return values should be stricter:
u8 is way more strict that number.


> there are 3 extra padding bits for each byte which makes no sense
There are 48 extra padding bits in each number.

junderw commented:
> Uint8Array is the closest thing to Uint5Array.

It also makes a clear statement to the API consumer that "this value should contain 8 bits". Uint8 literally says in it's name "consider this to be an 8 bit unsigned number. explicitly it is not 1 bit, or 4 bit, or 5 bit, it is 8 bit."

Crypto libraries accept Buffers and Uint8Arrays but not number[] for a reason, because they are dealing with bytes of data.

bech32 is dealing with arrays of 5-bit numbers. a 5-bit number is not a byte.

A number does not even *imply* anything about the number of bits within the number. The fact that it is implemented as a 64-bit double precision floating point is irrelevant. You choose number[] when you want an array of numbers, you choose Uint8Array when you want an array of 8 bit integers.

If you're not ok with 5-bit word arrays being returned as number[], then I won't force that decision onto this pull request. However, I also won't merge this. If you feel that strongly you can fork this library.

paulmillr commented:
I don't care much. I just want this to be merged in some way. It would be shitty if this ain't merged, since i've spent some time.

You can add a simple check if fromWords input is Uint8Array and return Uint8Array in this case.

junderw commented:
ok, I will review it with my team later this week (hopefully) and get some feedback from them as well.

As far as myself it looks good to me.

paulmillr commented:
Ping folks. Let's get this in! 😃

junderw commented:
Hey! Thanks again for the great work.

We haven't forgotten about this, we just keep getting a bunch of things popping up as high priority on our queue internally.

Perhaps @lukechilds or @fanatid want to give a review and also some insight as to how BitcoinJS (as a whole) should deal with Buffers moving forward.

It probably won't be too big of a deal just moving bech32 to Uint8Array and dealing with everything else later, but if we decide to handle things in some way after looking at the big picture, we might have to bump major again with bech32.

junderw commented on package.json:
```diff
@@ -20,20 +20,26 @@
   ],
   "license": "MIT",
-  "dependencies": {},
+  "dependencies": {
```
These 3 should be devDependencies

lukechilds commented:
>Perhaps @lukechilds or @fanatid want to give a review and also some insight as to how BitcoinJS (as a whole) should deal with Buffers moving forward.

I'm happy to help review this but I'm going to be busy until next week.

paulmillr commented:
@lukechilds ping?

That seems really simple.

junderw commented:
Yes, the change is really simple. Switch from Buffers to Uint8Array.

But this is a major change that affects tons of libraries that rely on this. Mainly bitcoinjs-lib.

Since tons of our other dependencies use Buffers, it would make more sense to start a larger project of moving things over to Uint8Array... perhaps even making a small util (much smaller than the Buffer class) to help with managing Uint8Arrays in upstream libraries.

To be honest, I think bitcoinjs-lib will rely more heavily on WASM and browser-based APIs (since modern browsers have gotten so much better over the years) moving forward.

But this is all a daunting task.

Myself, Luke, and other contributors have day jobs that don't really benefit from whether bitcoinjs-lib uses Buffers or not... so trying to find time to take on this task is difficult.

However, I do want to move forward with this.

One suggestion I have:

- Create a fork package on npm for people who want to use this now. (Maybe link to this issue in the README and deprecate the package once this is merged and published)

---

I am very interested to hear @fanatid opinion on migrating the whole bitcoinjs ecosystem away from Buffers.

Since Buffers are the main class that touches private key and signature data, it merits careful consideration. (Though I understand it is not that severe in the context of this bech32 library alone)

paulmillr commented:
The idea is simple. Switch to Uint8Arrays, which would allow the library to be used in browsers without deps. Which would potentially increase its usage.

If you don't want to maintain it, just keep using the old version. It's time to more forward with the progress here.

>so trying to find time to take on this task is difficult

Not really, it's just a question of priorities. I have very popular packages that require much more maintenance, and I have a bunch of businesses to run, and life outside the internet etc. Still finding time to work on stuff like this.

>affects tons of libraries that rely on this

If you follow semver and bump a major version, the users would not be affected.

>I do want to move forward with this

That seems like a "your call is important for us" I receive when I call bank's support & wait for 20 minutes.

>Create a fork package

I did this the second I published the rewrite. After that, decided to pull it back, because i've thought you're going to merge it in a couple days. The goal here is to let all bech32 users benefit from the change.
